### ISSUE TITLE
Error when invoking start_numpy_client() more than once

### ISSUE THREAD

**Issue by g-pichler** [adap/flower#1113]
### Describe the bug

When invoking start_numpy_client() more than once within the same python interperter, this will yield an error when the clients passed to it do not have a get_parameters() method.

The culprit seems to be line 168 of `src/py/flwr/client/app.py`:
```
   del NumPyClientWrapper.get_properties
```
If this deletion is invoked a second time, an AttributeError is raised. Probably there is a more graceful way of dealing with get_properties() inside NumPyClientWrapper, e.g., by returning an empty dictionary.

### Steps/Code to Reproduce

In `quickstart_pytorch` place clients.py with the following contents:
```
import time
from client import main as client_main

def main():
    for _ in range(2):
        time.sleep(5)
        client_main()

if __name__ == '__main__':
    main()
```
Then run the followin shell script:
```
#!/bin/bash

for i in `seq 0 1`; do
    echo "Starting client $i"
    python clients.py &
done

for i in `seq 0 1`; do
    echo "Starting Server $i ..."
    python server.py
done
```

### Expected Results

I expect two runs of training to complete within several minutes all processes should exit with exit code 0.

### Actual Results

The first training (and server process) completes without issues. In the second run, the clients fail with the following backtrace:
```
Traceback (most recent call last):
  File "./clients.py", line 12, in <module>
    main()
  File "./clients.py", line 8, in main
    client_main()
  File "./client.py", line 122, in main
    fl.client.start_numpy_client("[::]:8080", client=CifarClient())
  File "flwr/client/app.py", line 168, in start_numpy_client
    del NumPyClientWrapper.get_properties
AttributeError: get_properties
```

**Comment by danieljanes** [comment]
Thanks for the report @g-pichler. The current implementation is indeed not ideal, one additional drawback of the solution is a setup in which different client implementations - some with, some without implementing `get_properties` - are started in the same Python process. Because `get_properties` gets deleted from the class, it won't be available for those client that actually implement it.

The current solution was chosen over returning an empty dict (as in #1114) because a `NumPyClient` wrapped in a `NumPyClientWrapper` behaves exactly the same as / is indistinguishable from a regular `Client` that does not implement `get_properties`. Reducing the differences between a user-defined `Client` and a user-defined `NumPyClient` wrapped in a `NumPyClientWrapper` is important for consistency. The `start_client` implementation needs a way to check if `get_properties` was implemented *without* calling it (the function is user-provided, so it can potentially run for a long while).

Did you perhaps have alternative solutions in mind @g-pichler? CC @tanertopal

**Comment by g-pichler** [comment]
Yes, starting different clients (with and without get_properties) is certainly also impossible at the moment.

I can see, that simply returning an empty dict in the `NumPyClientWrapper` if `get_properties` is not defined is not nice as `code=OK` would not be correct in this case. How would you feel about the following idea:

- Make the default implementation of `get_properties()` in both `NumPyClient` and `Client` raise an `AttributeError`.
- Catch that `AttributeError` in `_get_properties()` in `message_handler.py` to generate an appropriate `PropertiesRes`.

Would that have the desired effect? I closed #1114 and created #1138, so you can take a look. I have to admit, I don't quite understand the need for the `start_client` implementation to check if `get_properties` was implemented without calling it. It does not seem to me that `start_client` checks this. How I see it, when the server sends an RPC for `get_properties`, there are 2 options for the client:

1. The client implements it, in which case it should be called (and might take a long time).
2. The client doesn't implement it and thus an `AttributeError(...)` is raised, caught and the appropriate message is returned to the server.

Probably I'm going wrong somewhere and there is something that I'm not thinking about/aware of.

PS: I don't think it's possible to run multiple clients simultaneously in threads. Is this something you are aware of? Should I open a dedicated Issue for this?

**Comment by danieljanes** [comment]
Sorry for the late reply @g-pichler, re-posting my reply from PR #1138 for future reference:

> We're currently reviewing the approach taken here to make all Client methods (not just get_properties) optional. Our progress so far:
>
> - We'll drop the current approach, one of the reasons is the bahviour you reported in the GitHub issue
> - We cannot just call the method and throw an exception from a default implementation (as it's implemented in this PR). We've briefly considered this approach, but one issue with it is that it does not allow us to quickly check if the method is implemented. Any approach where we have to call the method (to see if the user has implemented it) has the drawback that in the case where it is implemented, it can potentially take a long time before the method returns (and this call could itself raise an exception). If we want to build more infrastructure, for example, to automatically let the server know which client supports which method (this can vary across different clients in the same workload), we need a way to check if a method is implemented without having to call it.

Thanks for mentioning the issue about "multiple clients simultaneously in threads", would be great if you could open a dedicated issue for it to keep the discussion threads focused :)
